classes/file_storage.php

```diff
@@ -494,30 +494,70 @@ public function removeContentFile($file, $content) {
      *  Path to source directory
      * @param array $options
      *  For Moodle's file record
+     * @param \ZipArchive $archive
+     *  A currently open archive to store complete libraries. Used in recursive calls.
+     * @param string $relativepath
+     *  The current relative path from the base function call. Used in recursive calls.
+     *
      * @throws \Exception Unable to copy
      */
     // @codingStandardsIgnoreLine
-    private static function readFileTree($source, $options) {
+    private static function readFileTree($source, $options, $archive = null, $relativepath = '') {
         $dir = opendir($source);
         if ($dir === false) {
             trigger_error('Unable to open directory ' . $source, E_USER_WARNING);
             throw new \Exception('unabletocopy');
         }
 
+        // Create an empty zip to store a full lib archive as well.
+        if (empty($archive)) {
+            $archive = new \ZipArchive();
+            $path = tempnam(sys_get_temp_dir(),'libdir');
```

Comment:
this should probably be get_request_storage_directory() (which ends up in tmp but is configurable)

Reply:
this can probably stay as is for us